### PR TITLE
tpm2-util: auto-detect supported PCR banks

### PR DESCRIPTION
**Pull request by poettering**
Previously, we'd encode PCR policies strictly with the SHA256 PCR bank
set. However, as it appears not all hw impleent those. Sad.

Let's add some minimal logic to auto-detect supported PCR banks: if
SHA256 is supported, use that. But if not automatically fall back to
SHA1.

Fixes: #20134

### CODE REVIEW

**Comment by poettering** [comment]
This is not a complete fix, but preparation for one for now. We need to serialize in the LUKS superblock which bank we use, and this patch doesn't do that. 

Hence don't merge yet, I am mostly looking for feedback from @deify about this

**Comment by bluca** [comment]
Looks like `TPM2_ALG_SHA384` and `TPM2_ALG_SHA512` exist too, at least as definitions - what about matching on those too? For future-proofing

**Comment by poettering** [comment]
> Looks like `TPM2_ALG_SHA384` and `TPM2_ALG_SHA512` exist too, at least as definitions - what about matching on those too? For future-proofing

according to all docs I could find that's not what anyone uses or implements. Microsoft explicitly documents that it uses SHA256 if available, and SHA1 otherwise, so I think we should follow that. I mean, given that sha256 is safe and secure, and the one windows uses it's probably the best bet.

That said, the specs suggests SHA1 is actually optional, but SHA256 a requirement:

https://trustedcomputinggroup.org/wp-content/uploads/PC-Client-Specific-Platform-TPM-Profile-for-TPM-2p0-v1p05p_r14_pub.pdf

Confusing.

**Comment by bluca** [comment]
the tpm library defines them though?

https://sources.debian.org/src/tpm2-tss/3.0.3-2/include/tss2/tss2_tpm2_types.h/?hl=84#L84

And at least some projects like fwupd appear to expect them:

https://sources.debian.org/src/fwupd/1.5.7-4/plugins/tpm-eventlog/fu-tpm-eventlog-common.c/?hl=52#L52
https://codesearch.debian.net/search?q=TPM2_ALG_SHA512&literal=1

**Comment by poettering** [comment]
Well, of course, TPM2 devices support these hashes in general. But not sure any do for the PCR stuff. And even if they did there's the question if we want to use it then. I am not sure. SHA256 should be good enough for these usecases. SHA1 is typically not. Hence I'd make a choice here, and focus on SHA256, and provide a fallback if that is not available. But not sure why we should bother with SHA512 if a) no hw I know of supports it, b) windows doesn't use it, c) it gives us no real-life security benefit over sha256, since that should be good enough.

**Comment by bluca** [comment]
fair enough - was just thinking in term of future proofing as these things evolve, but we can worry about that later

**Comment by poettering** [comment]
should be ready for review/merge now.

**Comment by bluca** [comment]
Some missing ifdefs somewhere?

```
../../src/systemd/src/shared/tpm2-util.c:1069:29: error: use of undeclared identifier 'TPM2_ALG_SHA256'
        return IN_SET(bank, TPM2_ALG_SHA256, TPM2_ALG_SHA1);
                            ^
../../src/systemd/src/shared/tpm2-util.c:1069:46: error: use of undeclared identifier 'TPM2_ALG_SHA1'
        return IN_SET(bank, TPM2_ALG_SHA256, TPM2_ALG_SHA1);
                                             ^
../../src/systemd/src/shared/tpm2-util.c:1069:16: error: invalid application of 'sizeof' to an incomplete type 'const long double []'
        return IN_SET(bank, TPM2_ALG_SHA256, TPM2_ALG_SHA1);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/systemd/src/fundamental/macro-fundamental.h:200:27: note: expanded from macro 'IN_SET'
                assert_cc(ELEMENTSOF(__assert_in_set) <= 20); \
                ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/systemd/src/fundamental/macro-fundamental.h:30:23: note: expanded from macro 'ELEMENTSOF'
                sizeof(x)/sizeof((x)[0]),                               \
                      ^
../../src/systemd/src/fundamental/macro-fundamental.h:42:23: note: expanded from macro 'assert_cc'
        static_assert(expr, #expr)
        ~~~~~~~~~~~~~~^~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/sys/cdefs.h:407:53: note: expanded from macro '_Static_assert'
      [!!sizeof (struct { int __error_if_negative: (expr) ? 2 : -1; })]
                                                    ^~~~
../../src/systemd/src/shared/tpm2-util.c:1069:29: error: use of undeclared identifier 'TPM2_ALG_SHA256'
        return IN_SET(bank, TPM2_ALG_SHA256, TPM2_ALG_SHA1);
                            ^
../../src/systemd/src/shared/tpm2-util.c:1069:46: error: use of undeclared identifier 'TPM2_ALG_SHA1'
        return IN_SET(bank, TPM2_ALG_SHA256, TPM2_ALG_SHA1);
                                             ^
../../src/systemd/src/shared/tpm2-util.c:1074:21: error: use of undeclared identifier 'TPM2_ALG_SHA256'
        if (bank == TPM2_ALG_SHA256)
                    ^
../../src/systemd/src/shared/tpm2-util.c:1076:21: error: use of undeclared identifier 'TPM2_ALG_SHA1'
        if (bank == TPM2_ALG_SHA1)
                    ^
../../src/systemd/src/shared/tpm2-util.c:1083:24: error: use of undeclared identifier 'TPM2_ALG_SHA256'
                return TPM2_ALG_SHA256;
                       ^
../../src/systemd/src/shared/tpm2-util.c:1085:24: error: use of undeclared identifier 'TPM2_ALG_SHA1'
                return TPM2_ALG_SHA1;
                       ^
9 errors generated.
```

File: src/shared/creds-util.c [hunk 1]
```diff
@@ -739,6 +745,10 @@ int decrypt_credential_and_warn(
 
                 if (le64toh(t->pcr_mask) >= (UINT64_C(1) << TPM2_PCRS_MAX))
                         return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "TPM2 PCR mask out of range.");
+                if (!tpm2_pcr_bank_supported(le16toh(t->pcr_bank)))
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "TPM2 PCR bank invalid or not supported");
+                if (le16toh(t->_zero) != 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "TPM2 zero space not zero.");
```
Comment (bluca) [inline src/shared/creds-util.c]: /zero space/padding space/ to be clearer?

File: src/shared/tpm2-util.c [hunk 2]
```diff
@@ -310,11 +312,87 @@ static int tpm2_make_primary(
...
+
+        for (size_t i = 0; i < pcap->data.assignedPCR.count; i++) {
+                bool valid = true;
+
+                /* As per
+                 * https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf a
+                 * TPM2 on a Client PC must have at least 24 PCRs. If this TPM has less, just skip over
+                 * it. */
+                if (pcap->data.assignedPCR.pcrSelections[i].sizeofSelect < TPM2_PCRS_MAX/8)
+                        continue;
```
Comment (bluca) [inline src/shared/tpm2-util.c:351]: Should an error be logged (debug level) here (and below)? Otherwise it would be pretty confusing for a user, no?

**Comment by poettering** [comment]
force pushed new version with hopefully everything addressed. And I hope it passes CI this time (i added some compat ifdefs to make sure things should compile without tpm2 libs, as the CI apparently lacks them)

**Review by bluca — APPROVED** [review]
lgtm

**Pull request merged**